// @jscpd/finder/dist/reporters/console-full.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const reports_1 = require("../utils/reports");
const clone_found_1 = require("../utils/clone-found");
const safe_1 = require("colors/safe");
const Table = require('cli-table3');
const TABLE_OPTIONS = {
    chars: {
        top: '',
        'top-mid': '',
        'top-left': '',
        'top-right': '',
        bottom: '',
        'bottom-mid': '',
        'bottom-left': '',
        'bottom-right': '',
        left: '',
        'left-mid': '',
        mid: '',
        'mid-mid': '',
        right: '',
        'right-mid': '',
        middle: '│',
    },
};
class ConsoleFullReporter {
    constructor(options) {
        this.options = options;
    }
    report(clones) {
        clones.forEach((clone) => {
            this.cloneFullFound(clone);
        });
        console.log(safe_1.grey(`Found ${clones.length} clones.`));
    }
    cloneFullFound(clone) {
        const table = new Table(TABLE_OPTIONS);
        clone_found_1.cloneFound(clone, this.options);
        clone.duplicationA.fragment.split('\n').forEach((line, position) => {
            (table).push(reports_1.generateLine(clone, position, line));
        });
        console.log(table.toString());
        console.log('');
    }
}
exports.ConsoleFullReporter = ConsoleFullReporter;
