// @jscpd/finder/dist/reporters/json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_extra_1 = require("fs-extra");
const core_1 = require("@jscpd/core");
const reports_1 = require("../utils/reports");
const safe_1 = require("colors/safe");
const path_1 = require("path");
class JsonReporter {
    constructor(options) {
        this.options = options;
    }
    generateJson(clones, statistics) {
        return {
            statistics,
            duplicates: clones.map(clone => this.cloneFound(clone))
        };
    }
    report(clones, statistic) {
        const json = this.generateJson(clones, statistic);
        fs_extra_1.ensureDirSync(core_1.getOption('output', this.options));
        fs_extra_1.writeFileSync(core_1.getOption('output', this.options) + '/jscpd-report.json', JSON.stringify(json, null, '  '));
        console.log(safe_1.green(`JSON report saved to ${path_1.join(this.options.output, 'jscpd-report.json')}`));
    }
    cloneFound(clone) {
        const startLineA = clone.duplicationA.start.line;
        const endLineA = clone.duplicationA.end.line;
        const startLineB = clone.duplicationB.start.line;
        const endLineB = clone.duplicationB.end.line;
        return {
            format: clone.format,
            lines: endLineA - startLineA + 1,
            fragment: clone.duplicationA.fragment,
            tokens: 0,
            firstFile: {
                name: reports_1.getPath(clone.duplicationA.sourceId, this.options),
                start: startLineA,
                end: endLineA,
                startLoc: clone.duplicationA.start,
                endLoc: clone.duplicationA.end,
                blame: clone.duplicationA.blame,
            },
            secondFile: {
                name: reports_1.getPath(clone.duplicationB.sourceId, this.options),
                start: startLineB,
                end: endLineB,
                startLoc: clone.duplicationB.start,
                endLoc: clone.duplicationB.end,
                blame: clone.duplicationB.blame,
            },
        };
    }
}
exports.JsonReporter = JsonReporter;
