// @jscpd/core/dist/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mode_1 = require("./mode");
function getDefaultOptions() {
    return {
        executionId: new Date().toISOString(),
        path: [process.cwd()],
        mode: mode_1.getModeHandler('mild'),
        minLines: 5,
        maxLines: 1000,
        maxSize: '100kb',
        minTokens: 50,
        output: './report',
        reporters: ['console'],
        ignore: [],
        threshold: undefined,
        formatsExts: {},
        debug: false,
        silent: false,
        blame: false,
        cache: true,
        absolute: false,
        noSymlinks: false,
        skipLocal: false,
        ignoreCase: false,
        gitignore: false,
        reportersOptions: {},
        exitCode: 0,
    };
}
exports.getDefaultOptions = getDefaultOptions;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function getOption(name, options) {
    const defaultOptions = getDefaultOptions();
    return options ? options[name] || defaultOptions[name] : defaultOptions[name];
}
exports.getOption = getOption;
